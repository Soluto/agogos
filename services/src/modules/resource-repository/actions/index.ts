import applyResourceGroupUpdates from './update';
import applyResourceGroupDeletions from './delete';

export { applyResourceGroupUpdates, applyResourceGroupDeletions };
