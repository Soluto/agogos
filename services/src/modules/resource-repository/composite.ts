import {ResourceRepository, FetchLatestResult, ResourceGroup} from './types';
import {applyResourceGroupUpdates} from './updates';

export class CompositeResourceRepository implements ResourceRepository {
    constructor(protected repositories: ResourceRepository[]) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const results = await Promise.all(this.repositories.map(repo => repo.fetchLatest()));

        return results.reduce((res1, res2) => ({
            isNew: res1.isNew || res2.isNew,
            resourceGroup: applyResourceGroupUpdates(res1.resourceGroup, res2.resourceGroup),
        }));
    }

    getResourceGroup(): ResourceGroup {
        const rgs = this.repositories.map(r => r.getResourceGroup());

        return rgs.reduce((rg1, rg2) => applyResourceGroupUpdates(rg1, rg2));
    }

    async update(): Promise<void> {
        throw new Error('Multiplexed resource repository cannot handle updates');
    }

    async writePolicyAttachment(): Promise<void> {
        throw new Error('Multiplexed resource repository cannot handle updates');
    }

    public getPolicyAttachment(filename: string): Buffer {
        for (const repo of this.repositories) {
            const policyAttachment = repo.getPolicyAttachment(filename);
            if (policyAttachment) return policyAttachment;
        }

        throw new Error(`Policy attachment with the filename ${filename} was not found`);
    }

    public async initializePolicyAttachments() {
        await Promise.all(this.repositories.map(repo => repo.initializePolicyAttachments()));
    }
}
