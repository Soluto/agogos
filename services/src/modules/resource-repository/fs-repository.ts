import * as envVar from 'env-var';
import { FileSystemStorage } from '../storage';
import { ResourceRepository } from '.';

export class FileSystemResourceRepository extends ResourceRepository {
  protected policyAttachmentsFolderInitialized = false;

  constructor(
    protected storage: FileSystemStorage,
    protected resourceFilePath: string,
    protected policyAttachmentsFolderPath: string,
    protected isRegistry = false
  ) {
    super(storage, resourceFilePath, policyAttachmentsFolderPath, isRegistry);
  }

  async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
    await this.initializePolicyAttachmentsFolder();
    await super.writePolicyAttachment(filename, content);
  }

  async listPolicyAttachments() {
    await this.initializePolicyAttachmentsFolder();
    return super.listPolicyAttachments();
  }

  private async initializePolicyAttachmentsFolder() {
    if (this.policyAttachmentsFolderInitialized) return;

    await this.storage.mkdir(this.policyAttachmentsFolderPath, { recursive: true });
    this.policyAttachmentsFolderInitialized = true;
  }

  static fromEnvironment(options: { isRegistry: boolean } = { isRegistry: false }) {
    const resourceFilePath = envVar.get('FS_RESOURCE_REPOSITORY_PATH').required().asString();
    const policyAttachmentsFolderPath = envVar
      .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
      .required()
      .asString();

    const fsStorage = new FileSystemStorage();
    return new FileSystemResourceRepository(
      fsStorage,
      resourceFilePath,
      policyAttachmentsFolderPath,
      options.isRegistry
    );
  }
}
