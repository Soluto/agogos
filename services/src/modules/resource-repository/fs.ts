import {ResourceGroup, ResourceRepository} from '.';
import * as envVar from 'env-var';
import {promises as fs} from 'fs';
import * as path from 'path';
import pLimit from 'p-limit';
import * as config from '../config';
import logger from '../logger';
import {FetchLatestResult} from './types';

export class FileSystemResourceRepository implements ResourceRepository {
    protected current?: {mtime: number; rg: ResourceGroup};
    protected policyAttachmentsDirInitialized = false;
    protected policyAttachments: {[filename: string]: Buffer} = {};
    protected policyAttachmentsRefreshedAt?: Date;

    constructor(protected pathToFile: string, protected policyAttachmentsFolderPath: string) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const stats = await fs.stat(this.pathToFile);

        if (this.current && stats.mtimeMs === this.current.mtime) {
            return {isNew: false, resourceGroup: this.current.rg};
        }

        const contents = await fs.readFile(this.pathToFile, 'utf8');
        const rg = JSON.parse(contents) as ResourceGroup;
        this.current = {mtime: stats.mtimeMs, rg};

        return {isNew: true, resourceGroup: rg};
    }

    async update(rg: ResourceGroup): Promise<void> {
        await fs.writeFile(this.pathToFile, JSON.stringify(rg));
    }

    async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
        await this.initializePolicyAttachmentsDir();

        const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
        await fs.writeFile(filePath, content);
    }

    public getPolicyAttachment(filename: string): Buffer {
        return this.policyAttachments[filename];
    }

    public async initializePolicyAttachments() {
        try {
            await this.refreshPolicyAttachments();
        } catch (err) {
            logger.fatal({err}, 'Failed fetching fs policy attachments on startup');
            throw err;
        }

        setInterval(async () => {
            try {
                await this.refreshPolicyAttachments();
            } catch (err) {
                logger.error(
                    {err},
                    `Failed refreshing fs policy attachments, last successful refresh was at ${this.policyAttachmentsRefreshedAt}`
                );
            }
        }, config.resourceUpdateInterval);
    }

    private async refreshPolicyAttachments() {
        const newRefreshedAt = new Date();

        const allAttachments = await this.getPolicyAttachmentsList();
        const attachmentsToRefresh = allAttachments
            .filter(a => this.shouldRefreshPolicyAttachment(a))
            .map(a => a.filename);

        if (attachmentsToRefresh.length > 0) {
            const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);
            newAttachments.forEach(a => (this.policyAttachments[a.filename] = a.content));
        }

        this.policyAttachmentsRefreshedAt = newRefreshedAt;
    }

    private shouldRefreshPolicyAttachment({filename, updatedAt}: {filename: string; updatedAt: Date}) {
        if (!this.policyAttachments[filename]) return true;
        if (!this.policyAttachmentsRefreshedAt) return true;

        return updatedAt > this.policyAttachmentsRefreshedAt;
    }

    private async getPolicyAttachmentsList(): Promise<{filename: string; updatedAt: Date}[]> {
        const filenames = await fs.readdir(this.policyAttachmentsFolderPath);
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const stats = await fs.stat(filePath);

                    const updatedAt = stats.mtime;
                    return {filename, updatedAt};
                });
            })
        );
    }

    private async getPolicyAttachments(filenames: string[]): Promise<{filename: string; content: Buffer}[]> {
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const content = await fs.readFile(filePath);

                    return {filename, content};
                });
            })
        );
    }

    private async initializePolicyAttachmentsDir() {
        if (this.policyAttachmentsDirInitialized) return;

        await fs.mkdir(this.policyAttachmentsFolderPath, {recursive: true});
        this.policyAttachmentsDirInitialized = true;
    }

    static fromEnvironment() {
        const pathToFile = envVar
            .get('FS_RESOURCE_REPOSITORY_PATH')
            .required()
            .asString();
        const policyAttachmentsFolderPath = envVar
            .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
            .required()
            .asString();

        return new FileSystemResourceRepository(pathToFile, policyAttachmentsFolderPath);
    }
}
