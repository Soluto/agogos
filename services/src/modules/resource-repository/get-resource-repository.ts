import * as config from '../config';
import logger from '../logger';
import {
  CompositeResourceRepository,
  FileSystemResourceRepository,
  IResourceRepository,
  ResourceRepository,
  S3ResourceRepository,
} from '.';

export default function getResourceRepository(options?: { isRegistry: boolean }): IResourceRepository {
  const repositories: ResourceRepository[] = [];

  if (config.useS3ResourceRepository) {
    repositories.push(S3ResourceRepository.fromEnvironment(options));
  }

  if (config.useFileSystemResourceRepository) {
    repositories.push(FileSystemResourceRepository.fromEnvironment(options));
  }

  switch (repositories.length) {
    case 0:
      logger.fatal('Must enable at least one resource repository');
      throw new Error('Must enable at least one resource repository');
    case 1:
      return repositories[0];
    default:
      if (options?.isRegistry) {
        logger.fatal('Registry cannot have more than one resource repository');
        throw new Error('Registry cannot have more than one resource repository');
      }
      return new CompositeResourceRepository(repositories);
  }
}
