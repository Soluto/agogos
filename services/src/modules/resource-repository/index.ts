export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
