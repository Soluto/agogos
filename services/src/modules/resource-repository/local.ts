import {ResourceGroup, ResourceRepository} from '.';
import {localResourceRepositoryPath} from '../config';
import {promises as fs} from 'fs';

export class LocalResourceRepository implements ResourceRepository {
    protected localPath: string;

    constructor() {
        if (!localResourceRepositoryPath) {
            throw new Error("Must configure 'LOCAL_RESOURCE_REPO_PATH' for Local repository to work!");
        } else {
            this.localPath = localResourceRepositoryPath;
        }
    }

    /** Returns latest resource group */
    fetch(): Promise<ResourceGroup>;
    async fetch(): Promise<ResourceGroup | null> {
        const json = await fs.readFile(this.localPath, 'utf8');
        return JSON.parse(json);
    }

    async update(rg: ResourceGroup): Promise<void> {
        const {etag, ...rgWithoutEtag} = rg;
        await fs.writeFile(this.localPath, JSON.stringify(rgWithoutEtag));
    }
}
