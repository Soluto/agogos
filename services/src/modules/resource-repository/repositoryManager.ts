import {ResourceGroup, ResourceRepository, applyResourceGroupUpdates} from '.';
import {enableLocalResourceRepository, enableS3ResourceRepository} from '../config';
import {S3ResourceRepository} from './s3';
import {LocalResourceRepository} from './local';

function initRepos(): ResourceRepository[] {
    let repos = [];
    if (enableS3ResourceRepository) {
        repos.push(new S3ResourceRepository());
    }

    if (enableLocalResourceRepository) {
        repos.push(new LocalResourceRepository());
    }
    return repos;
}

export function fetchAll(): Promise<ResourceGroup>;
export function fetchAll(currentEtag?: string): Promise<ResourceGroup | null>;
export async function fetchAll(currentEtag?: string): Promise<ResourceGroup | null> {
    const repos = initRepos();
    const resources = await Promise.all(repos.map(repo => repo.fetch(currentEtag)));
    if (resources.length == 0) return null;
    let newRg: ResourceGroup = {schemas: [], upstreams: [], upstreamClientCredentials: []};

    for (const rg of resources) {
        if (rg != null) {
            newRg = applyResourceGroupUpdates(newRg, rg);
        }
    }

    return newRg;
}

export async function updateAll(rg: ResourceGroup): Promise<void> {
    const repos = initRepos();
    await Promise.all(repos.map(repo => repo.update(rg)));
}
