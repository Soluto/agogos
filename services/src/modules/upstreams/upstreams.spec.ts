import { FastifyRequest } from 'fastify';
import * as _ from 'lodash';
import { ResourceMetadata, ResourceGroup } from '../resource-repository';
import { applyUpstream, RequestParams } from '.';

interface TestCase {
  resourceGroupPart: Partial<Pick<ResourceGroup, 'upstreams' | 'defaultUpstream'>>;
  requestParams: Partial<RequestParams>;
  incomingRequest?: Pick<FastifyRequest, 'headers'>;
}

const metadata: ResourceMetadata = {
  namespace: 'upstream-test',
  name: 'upstream',
};

const testCases: [string, TestCase][] = [
  [
    'No upstream',
    {
      resourceGroupPart: {},
      requestParams: {},
    },
  ],
  [
    'Change url',
    {
      resourceGroupPart: {
        upstreams: [
          {
            metadata,
            sourceHosts: ['maps.google.com'],
            targetOrigin: 'https://www.facebook.com',
          },
        ],
      },
      requestParams: {},
    },
  ],
  [
    'Add header',
    {
      resourceGroupPart: {
        upstreams: [
          {
            metadata,
            sourceHosts: ['maps.google.com'],
            headers: [{ name: 'x-api-client', value: 'upstream-tests' }],
          },
        ],
      },
      requestParams: {},
    },
  ],
  [
    'Add header from incoming request',
    {
      resourceGroupPart: {
        upstreams: [
          {
            metadata,
            sourceHosts: ['maps.google.com'],
            headers: [{ name: 'transactionId', value: '{incomingRequest?.headers?.transactionId}' }],
          },
        ],
      },
      requestParams: {},
      incomingRequest: {
        headers: {
          transactionId: '1',
        },
      },
    },
  ],
  [
    'Default default upstream',
    {
      resourceGroupPart: {},
      requestParams: {},
      incomingRequest: {
        headers: {
          authorization: 'Bearer S0mTh1nG',
        },
      },
    },
  ],
  [
    'Custom default upstream',
    {
      resourceGroupPart: {
        defaultUpstream: {
          headers: [{ name: 'authorization', value: '{incomingRequest?.headers?.authorization.toLowerCase()}' }],
        },
      },
      requestParams: {},
      incomingRequest: {
        headers: {
          authorization: 'Bearer S0mTh1nG',
        },
      },
    },
  ],
  [
    'Custom default upstream and upstream',
    {
      resourceGroupPart: {
        upstreams: [
          {
            metadata,
            sourceHosts: ['maps.google.com'],
            headers: [{ name: 'authorization', value: '{incomingRequest?.headers?.authorization.toUpperCase()}' }],
          },
        ],
        defaultUpstream: {
          headers: [{ name: 'authorization', value: '{incomingRequest?.headers?.authorization.toLowerCase()}' }],
        },
      },
      requestParams: {},
      incomingRequest: {
        headers: {
          authorization: 'Bearer S0mTh1nG',
        },
      },
    },
  ],
];

describe('Upstreams', () => {
  const defaultResourceGroup: ResourceGroup = {
    schemas: [],
    upstreams: [],
    upstreamClientCredentials: [],
    policies: [],
  };

  const defaultRequestParams: RequestParams = {
    url: new URL('https://maps.google.com/api/v1/get-location?city=TelAviv'),
    headers: {
      ['Host']: 'maps.google.com',
      ['Content-Type']: 'application/json',
    },
    method: 'POST',
  };

  test.each(testCases)('%s', async (_testCaseName, { requestParams, resourceGroupPart, incomingRequest }) => {
    const resourceGroup = { ...defaultResourceGroup, ...resourceGroupPart };
    const request = _.cloneDeep({ ...defaultRequestParams, ...requestParams });
    const result = await applyUpstream(request, resourceGroup, undefined as any, incomingRequest);
    expect(result).toMatchSnapshot();
  });
});
