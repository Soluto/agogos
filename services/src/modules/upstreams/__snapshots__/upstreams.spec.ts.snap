// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upstreams Add header 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json",
    "Host": "maps.google.com",
    "x-api-client": "upstream-tests",
  },
  "method": "POST",
  "url": "https://maps.google.com/api/v1/get-location?city=TelAviv",
}
`;

exports[`Upstreams Add header from incoming request 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json",
    "Host": "maps.google.com",
    "transactionId": "1",
  },
  "method": "POST",
  "url": "https://maps.google.com/api/v1/get-location?city=TelAviv",
}
`;

exports[`Upstreams Change url 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json",
    "Host": "maps.google.com",
  },
  "method": "POST",
  "url": "https://www.facebook.com/api/v1/get-location?city=TelAviv",
}
`;

exports[`Upstreams Custom default upstream 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json",
    "Host": "maps.google.com",
    "authorization": "bearer s0mth1ng",
  },
  "method": "POST",
  "url": "https://maps.google.com/api/v1/get-location?city=TelAviv",
}
`;

exports[`Upstreams Custom default upstream and upstream 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json",
    "Host": "maps.google.com",
    "authorization": "BEARER S0MTH1NG",
  },
  "method": "POST",
  "url": "https://maps.google.com/api/v1/get-location?city=TelAviv",
}
`;

exports[`Upstreams Default default upstream 1`] = `
Object {
  "headers": Object {
    "Authorization": "Bearer S0mTh1nG",
    "Content-Type": "application/json",
    "Host": "maps.google.com",
  },
  "method": "POST",
  "url": "https://maps.google.com/api/v1/get-location?city=TelAviv",
}
`;

exports[`Upstreams No upstream 1`] = `
Object {
  "headers": Object {
    "Content-Type": "application/json",
    "Host": "maps.google.com",
  },
  "method": "POST",
  "url": "https://maps.google.com/api/v1/get-location?city=TelAviv",
}
`;
